core/src/main/scala/kafka/server/ActionQueue.scala

```diff
@@ -0,0 +1,46 @@
+/**
+  * Licensed to the Apache Software Foundation (ASF) under one or more
+  * contributor license agreements.  See the NOTICE file distributed with
+  * this work for additional information regarding copyright ownership.
+  * The ASF licenses this file to You under the Apache License, Version 2.0
+  * (the "License"); you may not use this file except in compliance with
+  * the License.  You may obtain a copy of the License at
+  *
+  *    http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+package kafka.server
+
+import java.util.concurrent.LinkedBlockingDeque
+
+/**
+ * This queue is used to collect actions which need to be executed later. One use case is that ReplicaManager#appendRecords
+ * produces record changes so we need to check and complete delayed requests. In order to avoid conflicting locking,
+ * we add those actions to this queue and then complete them at the end of KafkaApis.handle() or DelayedJoin.onExpiration.
+ */
+class ActionQueue {
+  private val queue = new LinkedBlockingDeque[() => Unit]()
```

Comment:
Why are we using a blocking queue? It doesn't seem like we ever need the blocking functionality. Am I missing something?

Reply:
you are right. How about using ```ConcurrentLinkedQueue``` instead?

Reply:
Yes, I think that's better. One thing I was wondering about is whether contention is going to be an issue for this `ActionQueue`. Multiple threads are adding items to it and then trying to consume from it. I haven't thought about all the details, but would a thread local work better? In that case, each thread would add and then drain. This works fine for the request threads, but I wasn't sure about the other case that @junrao pointed out.

Reply:
Handler (thread) can have local ActionQueue and it is passed to each method to collect delayed action. ```DelayedJoin``` is specific case since the delayed actions are possible to be access by two thread (timeout thread and handler thread). A simple way to address thread local is that ```DelayedJoin``` owns a ActionQueue and the queue is passed to ```coordinator.onCompleteJoin``` and then the queue is consumed by ```DelayedJoin.onExpiration```. Both ```onComplete``` and ```onExpiration``` are thread-safe so use a thread local queue is safe.

Reply:
Maybe we can go with a single `ActionQueue` in this PR. We can submit a separate PR for reducing the contention by having one per thread.

Reply:
Sure. I will file a PR to follow the pattern in #9229

In order to simplify code base, I will revert the action queue in "per server" :)